kupfer/plugin/id2.py

```diff
@@ -0,0 +1,135 @@
+    *[{"key": k[2:], 
+      "label": "URL for "+k[2:].upper(), 
+      "type": str, 
+      "value": id2data[k]["url"] } for k in SETTING_URL_KEYS if k in id2data]
+)
+
+# global table holding identifier classes and icon names
+CLASS_NAMES={
+    "i": ["Identifier", "edit-find"],
+    "g": ["Geo-Location", "applications-internet"],
+    "d": ["Document", "x-office-document"],
+    "o": ["Object", "package-x-generic"],
+    "h": ["Hash", "dialog-password"], 
+    "t": ["Time", "x-office-calendar"],
+    "w": ["Person", "stock_person" ],
+}
+
+
+
+def guessId(token):
+    """Test if "token" matches the regex of any identifier and return all found types."""
+    token = token.strip()
+    types = []
+    for key, entry in id2data.items():
+        if key in SETTING_URL_KEYS and len(__kupfer_settings__[key[2:]]) < 3: continue 
+        if len(token) in entry["lens"]:
+            match = entry["recomp"].match(token)
+            if match:
+                entry["part"]=match.group(1)
+                types.append(entry)
+    return types
+
+
+
+def parseIdentifierLengthsOnce():
+    """parse "len" key string and assign a list of integers to speed up execution"""
+    if "lens" in id2data["d:doi"].keys(): return # check if already done
+    lens = []
+    for key, entry in id2data.items():
+        parts = entry['len'].split(",")
+        for part in parts:
+            nums = part.split("-")
+            imin = int(nums[0])
+            if(len(nums) == 2):
+                if(nums[1] == ""): nums[1] = "40"
+                imax = int(nums[1])
+                lens = list(range(imin, imax+1))
+            else:
+                lens.append(imin)
+
+        id2data[key]['lens'] = lens
+
+
+
+def precompileRegexOnce():
+    if "recomp" in id2data["d:doi"].keys(): return # check if already done
+    for _, entry in id2data.items():
+        entry["recomp"] = re.compile(r'^'+entry["re"]+r'$')
+
+
+
+class LookupID (Action):
+    def __init__(self):
+        Action.__init__(self, name=_("Open in Browser"))
+        self.foundId = {}
+        self.rank_adjust = 3     # Rank our result slightly higher than default. Since we match a regex, we probably offer what the user is looking for.
+        self.icon_name = "web-browser"
+
+    def activate(self, leaf):
+        """Called when item is selected via Enter/click"""
+        if self.foundId["id"] in SETTING_URL_KEYS:
+            query_url = __kupfer_settings__[self.foundId["id"][2:]] + self.foundId["part"]
+        else:
+            query_url = self.foundId["url"] + self.foundId["part"]
+        utils.show_url(query_url)
+
+    def item_types(self):
+        yield TextLeaf
+
+    def valid_for_item(self, leaf):
+        if len(leaf.object) < 5: return False
+        types = guessId(leaf.object)
+        if not types: return False
+        self.foundId = types[0]
+        idclass = types[0]["id"][0]
+        leaf.get_description = lambda: (self.foundId["desc"] + " ({})".format(CLASS_NAMES[idclass][0]) )
```

Comment:
Would I need to create individual `TextLeaf` subclasses and individual `TextSource` that yield them? E.g. create `class LocationTextLeaf(TextLeaf)` and then have a `LocationTextSource(TextSource)` that yields them because it is still just text? If yes, I would need to create 7 such class pairs and that is too much overhead IMO. 
Thanks for your feedback and your help!

Reply:
Ok I have applied that